Categorization guide (#392)
columnflow/columnflow: docs/user_guide/building_blocks/categories.md

Work in Progess! This guide has not been finalized and the presented code has not yet been tested.

TODO: make the code testable

# Categories

In Columnflow, there are many tools to create a complex and flexible categorization of all analysed
events. This guide presents, how to implement a set of categories in Columnflow and presents an
showcases, how to use the resulting categories via the

> {py:class}`~columnflow.tasks.yields.CreateYieldsTable` task.

Author: I feel like it's much better to use the yields table as an example here, because you actually see the impact of your categorization on your final yields, which makes it easy to cross-check whether the categorization works as intended (e.g. sub-categories should sum up to the yield of an inclusive category). This could be missed quite easily when looking at plots instead.

Of course, this is not a replacement for documentation of the `CreateYieldsTable` and it should still (and hopefully is) mentioned in the guide that you can also use categories for e.g. plotting

## How to use categories?

Assuming you used the analysis template to setup your analysis, there are already two categories
included, named ```incl``` and ```2j``` included. To test, that the categories are properly implemented,
we can use the CreateYieldsTable task:
```shell
law run cf.CreateYieldsTable --version v1 \
    --calibrators example --selector example --producers category_ids \
    --processes tt,st --categories incl,2j
```

When all tasks are finished, this should create a table presenting the event yield of each process

> for all categories. To create new categories, we essentially need three ingredients:

Author: The weights producer does not yet exist, so right now it is not necessary. 
I have used the yields task a few times for my analysis and it worked there.

1. We need to define our categories as {external+order:py:class}`order.category.Category` instances in the config
```{literalinclude} ../../../analysis_templates/cms_minimal/__cf_module_name__/config/analysis___cf_short_name_lc__.py
:language: python
:start-at: add_category(
:end-at: )
```

> 2. We need to write a {py:class}`~columnflow.categorization.Categorizer` that defines, which events

Author: I have no preferences here. In any case, you need to add both such that your `category_ids` Producer is working as intended.

to select for each category. The name of the Categorizer needs to be the same as the "selection"
attribute of the category inst.
```{literalinclude} ../../../analysis_templates/cms_minimal/__cf_module_name__/categorization/example.py
:language: python
```
Keep in mind that the module in which your define your Categorizer needs to be included in the law config

3. We need to run the {py:class}`~columnflow.production.categories.category_ids` Producer to create
the ```category_ids``` column. This is directly included via the ```--producers category_ids```,
but you could also run the ```category_ids``` Producer as part of another Selector or Producer.
```python
events = self[category_ids](events, **kwargs)
```

:::{dropdown} But how does this actually work?
The `category_ids` Producer loads all category instances from your config. For each leaf category inst

> (which is the "smallest unit" of categories, more to that later),

Author: I kept the brackets (since the mentioning of the  **leaf categories** is not that eye-catching), but removed the "more to that later"

    return events, ak.num(events.Jet.pt, axis=1) >= 3

@categorizer(uses={"Muon.pt", "Electron.pt"})
def cat_1e(self: Categorizer, events: ak.Array, **kwargs) -> tuple[ak.Array, ak.Array]:
    return events, (ak.num(events.Muon.pt, axis=1) == 0 & ak.num(events.Electron.pt, axis=1) == 1)

@categorizer(uses={"Muon.pt", "Electron.pt"})
def cat_1mu(self: Categorizer, events: ak.Array, **kwargs) -> tuple[ak.Array, ak.Array]:
    return events, (ak.num(events.Muon.pt, axis=1) == 1 & ak.num(events.Electron.pt, axis=1) == 0)
```

This code snippet produces 4 new categories, which (after rerunning the ```category_ids```
producer and recreating all necessary histograms) can already be used:

```shell
law run cf.CreateYieldsTable --version v1 \
    --calibrators example --selector example --producer category_ids \
    --processes tt,st --categories "*"
```

As a next step, one might want to combine these categories to categorize based on the number of
jets and leptons simutaneously. You could simply create all combinations by hand. The following code
snippet shows how to combine the "2jet" and the "1lep" category:

```python
cat_2jet = config.get_category("2jet")

> cat_1lep = config.get_category("1lep")

Author: Good point, has been fixed :+1:

# add combined category as child to the existing categories
cat_1e__2jet = cat_2jet.add_category(
    name="1e__2jet",
    id=cat_2jet.id + cat_1e.id,
    # our `category_ids` producer can automatically combine multiple selections via logical 'and'
    selection=[cat_1e.selection, cat_2jet.selection],
    label="1 electron, 0 muons, >2 jets",

> )

Author: Which part of the syntax are you referring to? Looks correct to me?

smallest units of categories (commonly called "leaf categories") to build our parent categories. For
example, the `2jet` category consists of the leaf categories `1e__2jet` and `1mu__2jet`. To select
events corresponding to `2jet`, we will add all events with category ids corresponding to either
`1e__2jet` or `1mu__2jet`.

:::{note} We might add unexpected selections by combining categorization.

In this example, the `cat_1e` and `cat_1mu` Categorizer only select events with exactly one lepton
(either electron or muon). This means, that events with zero or multiple leptons will not be considered
in our categorization, event when building e.g. the `2jet` category.
:::

Let's test if our leaf categories are working as intended:

```shell
law run cf.CreateYieldsTable --version v1 \
    --calibrators example --selector example --producer category_ids \
    --processes tt,st --categories 1e,2jet,1e__2jet,1e__3jet,1mu__2jet,1mu__3jet
```

We should see that the yield of the `2jet` category is the same as the `1e__2jet` and `1mu__2jet`
categories combined. This is to be expected, since the `2jet` category will be built by combining
histograms from all of their leaf categories.

> :::{dropdown} Note! Take care to define your categories orthogonal!

Author: I don't really want to extend all of this by default because this contains a code example which nobody should use in their analysis. I changed the first part to a note and added the code example as a dropdown to the note

        selection=f"cat_{n_jet}jet",
        label=f"{n_jet} jets",
    )
for i, (ht_down, ht_up) in enumerate(ht_categories):
    config.add_category(
        id=1000 * i,
        name=f"ht_{ht_down}to{ht_up}",
        selection=f"cat_ht{ht_down}to{ht_up}",
        label=f"HT in ({ht_down}, {ht_up}]",
    )

#
# define Categorizer modules
#

for n_jet in jet_categories:
    @categorizer(name="cat_{n_jet}jet", uses={"Jet.pt"})
    def cat_n_jet(self: Categorizer, events: ak.Array, **kwargs) -> tuple[ak.Array, ak.Array]:
        return events, ak.num(events.Jet.pt, axis=1) == n_jet

for n_lep in (1, 2, 3):
    @categorizer(name="cat_{n_lep}lep", uses={"Electron.pt", "Muon.pt"})
    def cat_n_lep(self: Categorizer, events: ak.Array, **kwargs) -> tuple[ak.Array, ak.Array]:
        return events, (ak.num(events.Electron.pt, axis=1) + ak.num(events.Muon.pt, axis=1)) == n_lep

for ht_down, ht_up in ht_categories:
    @categorizer(name="cat_ht{ht_down}to{ht_up}", uses={"Jet.pt"})
    def cat_ht(self: Categorizer, events: ak.Array, **kwargs) -> tuple[ak.Array, ak.Array]:
        ht = ak.sum(events.Jet.pt, axis=1)
        return events, ((ht > ht_down) & (ht <= ht_up))
```

> This will only create the (5+3+3) parent categories. To create all possible combinations of

Author: No, at this stage we only created the parent categories. The Leafs are built below via the `create_category_combinations` functions

these categories, we can use the {py:func}`~columnflow.config_util.create_category_combinations`
function that will do most of the work for us:

```python
def name_fn(root_cats):
    """ define how to combine names for category combinations """
    cat_name = "__".join(cat.name for cat in root_cats.values())
    return cat_name

def kwargs_fn(root_cats):
    """ define how to combine category attributes for category combinations """
    kwargs = {
        "id": sum([c.id for c in root_cats.values()]),
        "label": ", ".join([c.name for c in root_cats.values()]),
        # optional: store the information on how the combined category has been created
        "aux": {
            "root_cats": {key: value.name for key, value in root_cats.items()},
        },
    }
    return kwargs

#
# define how to combine categories
#

> category_blocks = OrderedDict({

Author: I'm not sure, the `root_cats` is one dict per category (e.g. containing 1e, 2jet, ht400to600), presenting which parent categories have been used to build this specific category.
On the other hand, the `category_blocks` is a single dict containing all categories that have been used to create the full categorization.

    skip_existing=False,
)

# let's check what categories have been produced
print(f"{n_cats} have been created by the create_category_combinations function")
all_cats = [cat.name for cat, _, _ in config.walk_categories()]
print(f"List of all cateogries in our config: \n{all_cats}")
```

To test our final set of categories, we can call our `CreateYieldsTable` task again.

```shell
law run cf.CreateYieldsTable --version v1 \
    --calibrators example --selector example --producer category_ids \
    --processes tt,st --categories "*"
```

:::{note} This categorization is not inclusive!

Since we only store leaf category ids, all the events that are not considered by one of the category
blocks will not be considered at all. For this set of categories it means that we implicitly added cuts
on the number of jets (>= 2 and <= 6), the number of leptons (>= 1 and <= 3), and HT (<= 600) as soon
as we use one of these categories.

TODO: it might be more instructive to build this example such that our categorization is inclusive.
:::

## Extend your categorization as part of a Producer

Some of your categories might need some computationally expensive reconstructed observables

> and can therefore only be called after certain requirements are met, or be called as part of a Producer.
> This can be achieved by adding category instances as part of the `Producer.init`. Columns from
> custom requirements can be added to a Producer via the `Producer.requires` in combination
> with the `Producer.setup`.

Reviewer: ```suggestion
and can therefore only be called after certain requirements are met.
In these cases, it might be beneficial to first produce columns and then load these columns when necessary.
To streamline this, you can set these dependencies as part of the Producer instance that creates the category ids (i.e. via the `category_ids` producer).

To properly set this up, add the category instances as part of the `Producer.init`. Columns from
custom requirements can be added to a Producer via the `Producer.requires` in combination
with the `Producer.setup`.
```

selected_events = get_events_from_cateogries(events, config_inst.get_category("my_category"))

# it is also possible to pass a list of categories. In that case, all events
# that belong to one of the requested categories are selected
selected_events = get_events_from_cateogries(events, ["cat1", "cat2"], config_inst)
```

This function automatically consideres category ids of all leaf categories from the requested
categories. This is especially helpful when your category contains multiple leaf categories.

## Key points (TL;DR)

- Categories are {external+order:py:class}`order.category.Category` instances defined in the config.
- We can add categories to each other; the "smallest unit" of category is referred to as "leaf category".
- We build each category by combining all of it's leaf categories (both in columns and in histograms).
- Leaf categories need to define one or multiple {py:class}`~columnflow.categorization.Categorizer`,
which is a function that defines whether or not an event belongs in this category.
- The `category_ids` column is produced via the
{py:class}`~columnflow.production.categories.category_ids` Producer.

> - Make sure that for each category, all of it's leaf categories are defined orthogonal to prevent double counting.

Reviewer: ```suggestion
- Make sure that for each category, all of its leaf categories are defined orthogonal to prevent double counting.
```

- Groups of categories can be combined via {py:func}`~columnflow.config_util.create_category_combinations`.

> - Make sure that each group of categories is inclusive; otherwise, all combined categories will not be inclusive aswell.

Reviewer: Not sure if we actually need this point - sometimes, you don't want to be inclusive in the first place, right?

Author: That's correct, however, I think it's important to mention the point that the category combinations also have an impact on the parent categories (e.g. my 2jet category also categorizes based on the number of leptons because I added a layer of lepton categories that is not inclusive).

Alternative proposal:

```
- Combining of categories can impact your parent categories; this can be prevented by defining
your categories such that each group of categories is inclusive
```

Work in Progess! This guide has not been finalized and the presented code has not yet been tested.

TODO: make the code testable

# Categories

In columnflow, there are many tools to create a complex and flexible categorization of all analysed
events.
Generally, this categorization can be layered.
We refer to the smallest building block of these layers as **leaf categories**, which can subsequently
be either run individually or combined into more complex categories.
This guide presents how to implement a set of categories in columnflow and shows how to use the
resulting categories via the {py:class}`~columnflow.tasks.yields.CreateYieldTable` task.

## Create a category

If you used the analysis template to setup your analysis, there are already two categories
included, named ```incl``` and ```2j```.
To test that the categories are properly implemented,
we can use the CreateYieldTable task:
```shell
law run cf.CreateYieldTable --version v1 \

>     --calibrators example --selector example --producers category_ids \

Reviewer: Shouldn't this also be 

```suggestion
    --calibrators example --selector example --producers example \
```

?

Alternatively, we could also first call `ProduceColumns` if you want to highlight the `category_ids` producer, or simply mention it in the text (which you do later on in any case)
